spec/common/providers/identity.spec.ts

```diff
@@ -85,4 +90,857 @@ describe('identity', () => {
+          name: 'projects/project-id',
+        },
+        timestamp: new Date(1000).toUTCString(),
+        additionalUserInfo: {
+          providerId: 'password',
+          profile: rawUserInfo,
+          username: undefined,
+          isNewUser: false,
+        },
+        credential: {
+          claims: undefined,
+          idToken: 'ID_TOKEN',
+          accessToken: 'ACCESS_TOKEN',
+          refreshToken: 'REFRESH_TOKEN',
+          expirationTime: new Date(time + 3600 * 1000).toUTCString(),
+          secret: 'OAUTH_TOKEN_SECRET',
+          providerId: 'password',
+          signInMethod: 'password',
+        },
+        params: {},
+      };
+
+      expect(
+        identity.parseAuthEventContext(decodedJwt, 'project-id', time)
+      ).to.deep.equal(context);
+    });
+
+    it('should parse a beforeCreate event', () => {
+      const time = now.getTime();
+      // beforeCreate
+      const decodedJwt = {
+        aud: 'https://us-east1-project_id.cloudfunctions.net/beforeCreate',
+        exp: 60 * 60 + 1,
+        iat: 1,
+        iss: 'https://securetoken.google.com/project_id',
+        sub: 'abcdefghijklmnopqrstuvwxyz',
+        uid: 'abcdefghijklmnopqrstuvwxyz',
+        event_id: 'EVENT_ID',
+        event_type: 'beforeCreate',
+        ip_address: '1.2.3.4',
+        user_agent: 'USER_AGENT',
+        locale: 'en',
+        sign_in_method: 'oidc.provider',
+        tenant_id: 'TENANT_ID',
+        user_record: {
+          uid: 'abcdefghijklmnopqrstuvwxyz',
+          email: 'user@gmail.com',
+          email_verified: true,
+          display_name: 'John Doe',
+          phone_number: '+11234567890',
+          provider_data: [
+            {
+              provider_id: 'oidc.provider',
+              email: 'user@gmail.com',
+              uid: 'user@gmail.com',
+              display_name: 'John Doe',
+            },
+          ],
+          photo_url: 'https://lh3.googleusercontent.com/1234567890/photo.jpg',
+          tokens_valid_after_time: 1476136676,
+          metadata: {
+            last_sign_in_time: 1476235905,
+            creation_time: 1476136676,
+          },
+          custom_claims: {
+            admin: true,
+            group_id: 'group123',
+          },
+          tenant_id: 'TENANT_ID',
+        },
+        oauth_id_token: 'ID_TOKEN',
+        oauth_access_token: 'ACCESS_TOKEN',
+        oauth_refresh_token: 'REFRESH_TOKEN',
+        oauth_token_secret: 'OAUTH_TOKEN_SECRET',
+        oauth_expires_in: 3600,
+        raw_user_info: JSON.stringify(rawUserInfo),
+      };
+      const context = {
+        locale: 'en',
+        ipAddress: '1.2.3.4',
+        userAgent: 'USER_AGENT',
+        eventId: 'EVENT_ID',
+        eventType:
+          'providers/cloud.auth/eventTypes/user.beforeCreate:oidc.provider',
+        authType: 'USER',
+        resource: {
+          service: 'identitytoolkit.googleapis.com',
+          name: 'projects/project-id/tenants/TENANT_ID',
```

Comment:
Should we verify that GCIP is always going to have one handler for all of auth and not have a handler per tenant? I would assume that the latter would make sense, but they know their product better than us obv.

Reply:
[According to their public docs](https://cloud.google.com/identity-platform/docs/blocking-functions), functions are registered at the project level (one handler per project). If the customer wishes to filter on users registering/logging in under a specific tenant, they can filter on the `user.tenantId` field inside the function.